src/playlist-loader.js

```diff
@@ -379,57 +389,104 @@
+        endIndex = i - 1;
+        knownEnd = segment.start;
+        if (endIndex < 0) {
+          // Somehow the first segment claims to start *after* the time we are
+          // searching for so just return it
+          return 0;
+        }
+        break;
+      }
+      if (segment.end !== undefined && segment.end > time) {
+        if (segment.start !== undefined && segment.start < time) {
+          // we've found the target segment exactly
           return i;
         }
+        endIndex = i;
+        knownEnd = segment.end;
         break;
       }
     }
-    i++;
 
-    // 2) Walk forward, testing each segment to see if `time` falls within it
-    for (j = i; j < this.media_.segments.length; j++) {
-      segment = this.media_.segments[j];
-      time -= segment.duration || targetDuration;
+    if (startIndex !== undefined) {
+      // We have a known-start point that is before our desired time so
+      // walk from that point forwards
+      time = time - knownStart;
+      for (i = startIndex; i < (endIndex || numSegments); i++) {
+        segment = this.media_.segments[i];
+        time -= segment.duration || targetDuration;
+        if (time < 0) {
+          return i;
+        }
+      }
 
-      if (time < 0) {
-        return j;
+      if (i === endIndex) {
+        // We haven't found a segment but we did hit a known end point
+        // so fallback to "Algorithm Jon" (ie. interpolation)
+        return startIndex + Math.floor(
+          ((originalTime - knownStart) / (knownEnd - knownStart)) *
+          (endIndex - startIndex));
       }
 
-      // 2a) If we discover a segment that has timeline information
-      // before finding the result segment, the playlist information
-      // must have been inaccurate. Start a binary search for the
-      // segment which contains `time`. If the guess turns out to be
-      // incorrect, we'll have more info to work with next time.
-      if (segment.start !== undefined || segment.end !== undefined) {
-        return Math.floor((j - i) * 0.5);
+      // We _still_ haven't found a segment so load the last one
+      return lastSegment;
+    } else if (endIndex !== undefined) {
+      // We only have a known-end point that is after our desired time so
+      // walk from that point backwards
+      time = knownEnd - time;
+      for (i = endIndex; i >= 0; i--) {
+        segment = this.media_.segments[i];
+        time -= segment.duration || targetDuration;
+        if (time < 0) {
+          return i;
+        }
       }
+      // We haven't found a segment so load the first one
+      return 0;
+    } else {
+      // We known nothing so use "Algorithm A"
+      i = 0;
+      for (i = 0; i < numSegments; i++) {
+        segment = this.media_.segments[i];
+        time -= segment.duration || targetDuration;
+        if (time < 0) {
+          return i;
+        }
+      }
+      // We are out of possible candidates so load the last one
+      return lastSegment;
```

Comment:
You've switched back to clamping return values to valid indices in the segment array. At the time I tried this, I felt like failure to find an index was indicative of a problem elsewhere in the code and clamping would obscure that. You think that worry is unfounded?